File: docs/assets/js/ats.js
// ATS — Apollonian Time System v0.6 reference (browser).
// Epoch: 1969-07-20T00:00:00Z (start of Apollo 11 landing day).
// Bloc 5 = 12:00 UTC exactly. The landing itself (20:17:40 UTC) sits at Δ 0 . 8.4.5.6 (frac 0.84560…).
// Spec: /spec/manifesto.en.md (or .fr.md).
//
// v0.6+: adds Δ/Δd arithmetic per spec §11.4. JS uses Number (float64) so
// the duration arithmetic is precise to ~15 significant digits — fine for
// any practical day-scale computation but documented as a precision floor.
// The Python reference (code/ats.py) uses Decimal for full exactness.

(function (global) {
  'use strict';

  const ATS_EPOCH_MS = Date.parse('1969-07-20T00:00:00Z');
  const ATS_DECIMALS = 5;
  const ATS_SCALE = 100000;
  const MS_PER_DAY = 86400000;

  function pad(n, width) {
    return String(n).padStart(width, '0');
  }

  // Convert a millisecond timestamp (UTC epoch ms) to an ATS object.
  // Strict floor truncation per spec §6 — a unit is published only once
  // it has fully elapsed.
  function atsFromMs(ms) {
    const deltaMs = ms - ATS_EPOCH_MS;
    const sign = deltaMs >= 0 ? 'T+' : 'T-';
    const absMs = Math.abs(deltaMs);

    const intDays = Math.floor(absMs / MS_PER_DAY);
    const dayRemainderMs = absMs - intDays * MS_PER_DAY;

    // Fractional digits: scale to 5 places, floor.
    const frac = Math.floor((dayRemainderMs * ATS_SCALE) / MS_PER_DAY);

    const kilo = Math.floor(intDays / 1000);
    let rem = intDays % 1000;
    const hecto = Math.floor(rem / 100);
    rem = rem % 100;
    const deka = Math.floor(rem / 10);
    const kin = rem % 10;

    return { sign, kilo, hecto, deka, kin, frac };
  }

  function toCanonical(ats) {
    return `${ats.sign} Δ ${ats.kilo}.${ats.hecto}.${ats.deka}.${ats.kin}.${pad(ats.frac, ATS_DECIMALS)}`;
  }

  // Short form: ΔK.H.D.Kin-BC.M  (no space, Kin always shown, 2-digit Bloc+Centi
  // pair, single Milli digit always emitted, sign assumed T+).
  function toShort(ats) {
    const bc = Math.floor(ats.frac / Math.pow(10, ATS_DECIMALS - 2));
    const milli = Math.floor(ats.frac / Math.pow(10, ATS_DECIMALS - 3)) % 10;
    return `Δ${ats.kilo}.${ats.hecto}.${ats.deka}.${ats.kin}-${pad(bc, 2)}.${milli}`;
  }

  // Decode a canonical ATS string back to milliseconds UTC.
  // Returns null when the string is not canonical.
  function canonicalToMs(str) {
    const m = /^\s*(T[+-])\s*Δ\s*(\d+)\.(\d)\.(\d)\.(\d)\.(\d{1,})\s*$/.exec(str);
    if (!m) return null;
    const sign = m[1];
    const kilo = parseInt(m[2], 10);
    const hecto = parseInt(m[3], 10);
    const deka = parseInt(m[4], 10);
    const kin = parseInt(m[5], 10);
    let fracStr = m[6].slice(0, ATS_DECIMALS).padEnd(ATS_DECIMALS, '0');
    const frac = parseInt(fracStr, 10);

    const days = kilo * 1000 + hecto * 100 + deka * 10 + kin + frac / ATS_SCALE;
    const deltaMs = (sign === 'T+' ? 1 : -1) * days * MS_PER_DAY;
    return ATS_EPOCH_MS + deltaMs;
  }

  // -- v0.6 arithmetic (spec §11.4) -------------------------------------

  function signedDays(ats) {
    // ats: {sign, kilo, hecto, deka, kin, frac}
    const days = ats.kilo * 1000 + ats.hecto * 100 + ats.deka * 10 + ats.kin
      + ats.frac / ATS_SCALE;
    return ats.sign === 'T-' ? -days : days;
  }

  function fromSignedDays(d) {
    const sign = d >= 0 ? 'T+' : 'T-';
    const abs = Math.abs(d);
    const intDays = Math.floor(abs);
    const frac = Math.floor((abs - intDays) * ATS_SCALE);
    const kilo = Math.floor(intDays / 1000);
    let rem = intDays % 1000;
    const hecto = Math.floor(rem / 100);
    rem = rem % 100;
    const deka = Math.floor(rem / 10);
    const kin = rem % 10;
    return { sign, kilo, hecto, deka, kin, frac };
  }

  function dur(signed) {
    // Plain shape so callers can JSON.stringify a duration.
    return { kind: 'dur', signedDays: Number(signed) };
  }

  function isDur(x) { return x && x.kind === 'dur'; }
  function isAts(x) { return x && typeof x.kilo === 'number' && typeof x.frac === 'number' && !x.kind; }

  function durToCanonical(d) {
    const sign = d.signedDays < 0 ? 'T-' : 'T+';
    const abs = Math.abs(d.signedDays);
    const intDays = Math.floor(abs);
    const frac = Math.floor((abs - intDays) * ATS_SCALE);
    const kilo = Math.floor(intDays / 1000);
    let rem = intDays % 1000;
    const hecto = Math.floor(rem / 100);
    rem = rem % 100;
    const deka = Math.floor(rem / 10);
    const kin = rem % 10;
    return `${sign} Δd ${kilo}.${hecto}.${deka}.${kin}.${pad(frac, ATS_DECIMALS)}`;
  }

  function durFromCanonical(str) {
    const m = /^\s*(T[+-])\s*Δd\s+(\d+)\.(\d)\.(\d)\.(\d)\.(\d+)\s*$/.exec(str);
    if (!m) return null;
    const sign = m[1];
    const kilo = parseInt(m[2], 10);
    const hecto = parseInt(m[3], 10);
    const deka = parseInt(m[4], 10);
    const kin = parseInt(m[5], 10);
    const fracStr = m[6].slice(0, ATS_DECIMALS).padEnd(ATS_DECIMALS, '0');
    const frac = parseInt(fracStr, 10);
    const days = kilo * 1000 + hecto * 100 + deka * 10 + kin + frac / ATS_SCALE;
    return dur(sign === 'T-' ? -days : days);
  }

  // Polymorphic add: (ats, dur) → ats ; (dur, dur) → dur.
  function add(a, b) {
    if (isAts(a) && isDur(b)) return fromSignedDays(signedDays(a) + b.signedDays);
    if (isDur(a) && isAts(b)) return fromSignedDays(a.signedDays + signedDays(b));
    if (isDur(a) && isDur(b)) return dur(a.signedDays + b.signedDays);
    throw new TypeError('ATS.add: unsupported operands');
  }

  // Polymorphic sub: (ats, ats) → dur ; (ats, dur) → ats ; (dur, dur) → dur.
  function sub(a, b) {
    if (isAts(a) && isAts(b)) return dur(signedDays(a) - signedDays(b));
    if (isAts(a) && isDur(b)) return fromSignedDays(signedDays(a) - b.signedDays);
    if (isDur(a) && isDur(b)) return dur(a.signedDays - b.signedDays);
    throw new TypeError('ATS.sub: unsupported operands');
  }

  function mul(d, n) {
    if (!isDur(d) || typeof n !== 'number') {
      throw new TypeError('ATS.mul expects (dur, number)');
    }
    return dur(d.signedDays * n);
  }

  function div(d, n) {
    if (!isDur(d) || typeof n !== 'number') {
      throw new TypeError('ATS.div expects (dur, number)');
    }
    return dur(d.signedDays / n);
  }

  function neg(d) { return dur(-d.signedDays); }
  function absDur(d) { return dur(Math.abs(d.signedDays)); }

  function cmp(a, b) {
    const sa = isDur(a) ? a.signedDays : signedDays(a);
    const sb = isDur(b) ? b.signedDays : signedDays(b);
    return sa < sb ? -1 : sa > sb ? 1 : 0;
  }
  function lt(a, b) { return cmp(a, b) <  0; }
  function le(a, b) { return cmp(a, b) <= 0; }
  function eq(a, b) { return cmp(a, b) === 0; }
  function gt(a, b) { return cmp(a, b) >  0; }
  function ge(a, b) { return cmp(a, b) >= 0; }

  global.ATS = {
    EPOCH_MS: ATS_EPOCH_MS,
    EPOCH_ISO: '1969-07-20T00:00:00Z',
    DECIMALS: ATS_DECIMALS,
    atsFromMs,
    toCanonical,
    toShort,
    canonicalToMs,
    // v0.6 — Δ/Δd algebra (spec §11.4)
    signedDays,
    fromSignedDays,
    dur,
    durToCanonical,
    durFromCanonical,
    add,
    sub,
    mul,
    div,
    neg,
    abs: absDur,
    cmp,
    lt, le, eq, gt, ge,
  };
})(typeof window !== 'undefined' ? window : globalThis);

File: code/ats.py
"""
ATS — Apollonian Time System v0.6 reference implementation.

Spec: ../spec/manifesto.en.md
Epoch: 1969-07-20T00:00:00Z (start of the Apollo 11 landing day, UTC).
The landing itself (1969-07-20T20:17:40Z) is a notable instant within Δ 0,
located at Bloc 8 / Centi 4 / Deka 5 / Kin 6 (frac 0.84560185…).
This alignment makes Bloc 5 = 12:00 UTC exactly (5 × 2h24m).

Canonical format: "T+ Δ K.H.D.Kin.fffff"
  - K (Kilo) is an unbounded non-negative integer.
  - H, D, Kin are digits 0..9 (Hecto, Deka, Kin).
  - fffff is 5 fractional digits (default precision; variable).

Short format (UI): "Δ K.H.D.Kin/cc"
  - K, H, D, Kin are Kilo (unbounded), Hecto, Deka, Kin.
  - Kin is always shown (even when zero) to keep the calendar reference
    unambiguous. No spaces around `/`.
  - cc is two fractional digits (Bloc + Centi).
  - Decoding the short form is intentionally lossy: lower fractional
    digits (Milli/Beat/Blink) assumed 0; sign assumed T+. Parsers accept
    optional whitespace around `/`.

Duration type Δd (v0.6+, spec §11.4):
  ATSDuration represents a signed amount of days as a Decimal, independent
  of any anchor. Use ``Δ - Δ → Δd`` for instant differences, ``Δ + Δd → Δ``
  to advance an instant, ``Δd ± Δd``, ``Δd * n``, ``Δd / n`` for plain
  duration algebra. Canonical form: ``T± Δd K.H.D.Kin.fffff`` (sign carried
  explicitly, K unbounded, frac floored to ATS_DECIMALS digits).

Rounding policy (spec §6): strict floor truncation (ROUND_FLOOR) is used
when reducing precision for display. ATS is a counter of completed
units; a digit is shown only once its unit has fully elapsed. The
internal Decimal counter is always exact.

Leap second policy (spec §8): aligned to POSIX — a day is 86 400 seconds.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from decimal import ROUND_FLOOR, Decimal
from zoneinfo import ZoneInfo

ATS_SYMBOL = "Δ"
ATS_DECIMALS = 5
ATS_SCALE = 10 ** ATS_DECIMALS
ATS_EPOCH = datetime(1969, 7, 20, 0, 0, 0, tzinfo=timezone.utc)

_US_PER_DAY = 86_400_000_000


ATS_DURATION_SYMBOL = "Δd"


@dataclass(frozen=True, eq=False)
class ATSDateTime:
    """ATS instant.

    sign:  "T+" or "T-"
    kilo:  unbounded non-negative integer
    hecto, deka, kin: digits 0..9
    frac:  0..ATS_SCALE-1 (5 default digits of day-fraction)
    """

    sign: str
    kilo: int
    hecto: int
    deka: int
    kin: int
    frac: int = 0

    def __post_init__(self) -> None:
        if self.sign not in ("T+", "T-"):
            raise ValueError("sign must be 'T+' or 'T-'")
        if self.kilo < 0:
            raise ValueError("kilo must be >= 0")
        for name, value in (("hecto", self.hecto), ("deka", self.deka), ("kin", self.kin)):
            if not 0 <= value <= 9:
                raise ValueError(f"{name} must be a digit 0..9")
        if not 0 <= self.frac < ATS_SCALE:
            raise ValueError(f"frac must be in 0..{ATS_SCALE - 1}")

    @property
    def integer_days(self) -> int:
        return self.kilo * 1_000 + self.hecto * 100 + self.deka * 10 + self.kin

    @property
    def total_days_decimal(self) -> Decimal:
        return Decimal(self.integer_days) + Decimal(self.frac) / Decimal(ATS_SCALE)

    def _signed_decimal_days(self) -> Decimal:
        """Total days as a signed Decimal (negative when sign == 'T-').

        T+ 0 and T- 0 collapse to the same Decimal(0), so equality across
        signs at the epoch behaves correctly.
        """
        d = self.total_days_decimal
        return -d if self.sign == "T-" else d

    def to_canonical(self) -> str:
        return (
            f"{self.sign} {ATS_SYMBOL} "
            f"{self.kilo}.{self.hecto}.{self.deka}.{self.kin}."
            f"{self.frac:0{ATS_DECIMALS}d}"
        )

    def to_short(self) -> str:
        """Conversational form: Δ K.H.D.Kin/cc

        Drops the sign and the lower fractional digits (Milli/Beat/Blink).
        Kin is always shown — even when zero — to keep the calendar
        reference unambiguous. No spaces around `/`.
        """
        cc = self.frac // (10 ** (ATS_DECIMALS - 2))
        return f"{ATS_SYMBOL} {self.kilo}.{self.hecto}.{self.deka}.{self.kin}/{cc:02d}"

    def __str__(self) -> str:
        return self.to_canonical()

    # -- Arithmetic (spec §11.4) -----------------------------------------

    def __add__(self, other: object) -> ATSDateTime:
        if isinstance(other, ATSDuration):
            return _ats_from_signed_days(self._signed_decimal_days() + other.signed_days)
        return NotImplemented

    def __radd__(self, other: object) -> ATSDateTime:
        # Allows ATSDuration + ATSDateTime; the duration's __add__ returns
        # NotImplemented for the (Δd, Δ) case so Python falls back here.
        return self.__add__(other)

    def __sub__(self, other: object):
        if isinstance(other, ATSDateTime):
            return ATSDuration(self._signed_decimal_days() - other._signed_decimal_days())
        if isinstance(other, ATSDuration):
            return _ats_from_signed_days(self._signed_decimal_days() - other.signed_days)
        return NotImplemented

    def __eq__(self, other: object) -> bool:
        if isinstance(other, ATSDateTime):
            return self._signed_decimal_days() == other._signed_decimal_days()
        return NotImplemented

    def __lt__(self, other: object) -> bool:
        if isinstance(other, ATSDateTime):
            return self._signed_decimal_days() < other._signed_decimal_days()
        return NotImplemented

    def __le__(self, other: object) -> bool:
        if isinstance(other, ATSDateTime):
            return self._signed_decimal_days() <= other._signed_decimal_days()
        return NotImplemented

    def __gt__(self, other: object) -> bool:
        if isinstance(other, ATSDateTime):
            return self._signed_decimal_days() > other._signed_decimal_days()
        return NotImplemented

    def __ge__(self, other: object) -> bool:
        if isinstance(other, ATSDateTime):
            return self._signed_decimal_days() >= other._signed_decimal_days()
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self._signed_decimal_days())


@dataclass(frozen=True, eq=False)
class ATSDuration:
    """ATS signed duration Δd, spec §11.4 (v0.6+).

    Stores the signed amount of days as an arbitrary-precision Decimal.
    Independent of any anchor — pure quantity. Display uses
    ``T± Δd K.H.D.Kin.fffff`` with the same floor-truncation policy as
    ATSDateTime (spec §6).
    """

    signed_days: Decimal = Decimal(0)

    def __post_init__(self) -> None:
        # Coerce to Decimal so callers may pass int / str / float without
        # surprises. frozen dataclasses require object.__setattr__ to mutate.
        if not isinstance(self.signed_days, Decimal):
            object.__setattr__(self, "signed_days", Decimal(self.signed_days))

    @classmethod
    def zero(cls) -> ATSDuration:
        return cls(Decimal(0))

    @property
    def sign(self) -> str:
        return "T-" if self.signed_days < 0 else "T+"

    @property
    def abs_days(self) -> Decimal:
        return -self.signed_days if self.signed_days < 0 else self.signed_days

    def to_canonical(self) -> str:
        integer_days, frac_int = _split_abs_days_floor(self.abs_days)
        kilo, hecto, deka, kin = _integer_days_to_places(integer_days)
        return (
            f"{self.sign} {ATS_DURATION_SYMBOL} "
            f"{kilo}.{hecto}.{deka}.{kin}."
            f"{frac_int:0{ATS_DECIMALS}d}"
        )

    def __str__(self) -> str:
        return self.to_canonical()

    def __add__(self, other: object) -> ATSDuration:
        if isinstance(other, ATSDuration):
            return ATSDuration(self.signed_days + other.signed_days)
        # Δd + Δ delegates to ATSDateTime.__radd__.
        return NotImplemented

    def __sub__(self, other: object) -> ATSDuration:
        if isinstance(other, ATSDuration):
            return ATSDuration(self.signed_days - other.signed_days)
        return NotImplemented

    def __mul__(self, n: object) -> ATSDuration:
        if isinstance(n, (int, Decimal)):
            return ATSDuration(self.signed_days * Decimal(n))
        return NotImplemented

    __rmul__ = __mul__

    def __truediv__(self, n: object) -> ATSDuration:
        if isinstance(n, (int, Decimal)):
            return ATSDuration(self.signed_days / Decimal(n))
        return NotImplemented

    def __neg__(self) -> ATSDuration:
        return ATSDuration(-self.signed_days)

    def __abs__(self) -> ATSDuration:
        return ATSDuration(self.abs_days)

    def __eq__(self, other: object) -> bool:
        if isinstance(other, ATSDuration):
            return self.signed_days == other.signed_days
        return NotImplemented

    def __lt__(self, other: object) -> bool:
        if isinstance(other, ATSDuration):
            return self.signed_days < other.signed_days
        return NotImplemented

    def __le__(self, other: object) -> bool:
        if isinstance(other, ATSDuration):
            return self.signed_days <= other.signed_days
        return NotImplemented

    def __gt__(self, other: object) -> bool:
        if isinstance(other, ATSDuration):
            return self.signed_days > other.signed_days
        return NotImplemented

    def __ge__(self, other: object) -> bool:
        if isinstance(other, ATSDuration):
            return self.signed_days >= other.signed_days
        return NotImplemented

    def __hash__(self) -> int:
        return hash(self.signed_days)


def _ats_from_signed_days(d: Decimal) -> ATSDateTime:
    """Build an ATSDateTime from a signed Decimal day count.

    Floor truncation is applied to the absolute value (spec §6).
    """
    sign = "T-" if d < 0 else "T+"
    abs_d = -d if d < 0 else d
    integer_days, frac_int = _split_abs_days_floor(abs_d)
    kilo, hecto, deka, kin = _integer_days_to_places(integer_days)
    return ATSDateTime(
        sign=sign,
        kilo=kilo,
        hecto=hecto,
        deka=deka,
        kin=kin,
        frac=frac_int,
    )


def _timedelta_to_decimal_days(td: timedelta) -> Decimal:
    total_us = td.days * _US_PER_DAY + td.seconds * 1_000_000 + td.microseconds
    return Decimal(total_us) / Decimal(_US_PER_DAY)


def _decimal_days_to_timedelta(days: Decimal) -> timedelta:
    total_us = (days * Decimal(_US_PER_DAY)).to_integral_value(rounding=ROUND_FLOOR)
    return timedelta(microseconds=int(total_us))


def _split_abs_days_floor(abs_days: Decimal) -> tuple[int, int]:
    """Split a non-negative day count into (integer_days, frac_digits).

    Uses strict floor truncation. A digit is published only after its
    unit has fully elapsed (no anticipation, no carry into a future day).
    """
    integer_days = int(abs_days.to_integral_value(rounding=ROUND_FLOOR))
    frac = abs_days - Decimal(integer_days)
    frac_int = int((frac * Decimal(ATS_SCALE)).to_integral_value(rounding=ROUND_FLOOR))
    return integer_days, frac_int


def _integer_days_to_places(integer_days: int) -> tuple[int, int, int, int]:
    kilo, rem = divmod(integer_days, 1_000)
    hecto, rem = divmod(rem, 100)
    deka, kin = divmod(rem, 10)
    return kilo, hecto, deka, kin


def gregorian_to_ats(dt: datetime, *, assume_tz: str = "UTC") -> ATSDateTime:
    """Convert a Gregorian datetime to ATS.

    Naive datetimes are interpreted in `assume_tz` (default UTC).
    """
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=ZoneInfo(assume_tz))
    dt_utc = dt.astimezone(timezone.utc)

    delta = dt_utc - ATS_EPOCH
    days = _timedelta_to_decimal_days(delta)

    sign = "T+" if days >= 0 else "T-"
    abs_days = -days if days < 0 else days

    integer_days, frac_int = _split_abs_days_floor(abs_days)
    kilo, hecto, deka, kin = _integer_days_to_places(integer_days)

    return ATSDateTime(sign=sign, kilo=kilo, hecto=hecto, deka=deka, kin=kin, frac=frac_int)


_ATS_CANON_RE = re.compile(
    r"^\s*(T[+-])\s*Δ\s*(\d+)\.(\d)\.(\d)\.(\d)\.(\d{1,})\s*$"
)

_ATS_SHORT_RE = re.compile(
    r"^\s*Δ\s*(\d+)\.(\d)\.(\d)\.(\d)\s*/\s*(\d{1,2})\s*$"
)


def ats_to_gregorian(ats: str, *, out_tz: str = "UTC", allow_short: bool = False) -> datetime:
    """Convert an ATS string back to a Gregorian datetime.

    Accepts the canonical form: "T+ Δ 20.7.8.2.50000"

    The short form ("Δ 20.7.8.2/50") is only accepted when
    ``allow_short=True``, because it is intentionally lossy
    (lower fractional digits zero-padded, sign assumed T+).
    Whitespace around the `/` is tolerated on input.
    """
    m = _ATS_CANON_RE.match(ats)
    if m:
        sign, kilo_s, hecto_s, deka_s, kin_s, frac_s = m.groups()
        frac_s = frac_s[:ATS_DECIMALS].ljust(ATS_DECIMALS, "0")
        ats_obj = ATSDateTime(
            sign=sign,
            kilo=int(kilo_s),
            hecto=int(hecto_s),
            deka=int(deka_s),
            kin=int(kin_s),
            frac=int(frac_s),
        )
        abs_days = ats_obj.total_days_decimal
        days = abs_days if sign == "T+" else -abs_days
        return (ATS_EPOCH + _decimal_days_to_timedelta(days)).astimezone(ZoneInfo(out_tz))

    m = _ATS_SHORT_RE.match(ats)
    if m:
        if not allow_short:
            raise ValueError(
                "Refusing to decode short ATS form without explicit opt-in "
                "(pass allow_short=True). Short form is lossy: lower "
                "fractional digits zero-padded, sign assumed T+."
            )
        kilo_s, hecto_s, deka_s, kin_s, cc_s = m.groups()
        cc = int(cc_s)
        if not 0 <= cc <= 99:
            raise ValueError("cc must be in 0..99")
        frac = cc * (10 ** (ATS_DECIMALS - 2))
        ats_obj = ATSDateTime(
            sign="T+",
            kilo=int(kilo_s),
            hecto=int(hecto_s),
            deka=int(deka_s),
            kin=int(kin_s),
            frac=frac,
        )
        abs_days = ats_obj.total_days_decimal
        return (ATS_EPOCH + _decimal_days_to_timedelta(abs_days)).astimezone(ZoneInfo(out_tz))

    raise ValueError(f"Invalid ATS format: {ats!r}")


if __name__ == "__main__":
    now_utc = datetime.now(timezone.utc)
    ats_now = gregorian_to_ats(now_utc)
    print("Gregorian (UTC):", now_utc.isoformat())
    print("ATS canonical :", ats_now.to_canonical())
    print("ATS short     :", ats_now.to_short())

    back = ats_to_gregorian(ats_now.to_canonical(), out_tz="UTC")
    print("Round-trip UTC:", back.isoformat())
    print("Drift (us)    :", abs((now_utc - back).total_seconds()) * 1e6)

    # v0.6: Δd arithmetic demo (spec §11.4)
    one_day = ATSDuration(Decimal(1))
    print("Δd one day     :", one_day.to_canonical())
    tomorrow = ats_now + one_day
    print("Δ tomorrow     :", tomorrow.to_canonical())
    delta = tomorrow - ats_now
    print("Δd round-trip  :", delta.to_canonical(), "(== 1 day:", delta == one_day, ")")
